src/lib.rs

```diff
@@ -104,22 +106,29 @@ pub mod flash {
         ///   - DMA must not access flash memory
         ///
         /// `addr` and `len` parameters must be valid and are not checked.
-        pub unsafe fn flash_range_program(&self, addr: u32, data: &[u8]) {
+        pub unsafe fn flash_range_program(&mut self, addr: u32, data: &[u8]) {
             self.write_flash_inner(addr, DataOrLen::Data(data), false);
         }
 
-        unsafe fn write_flash_inner(&self, addr: u32, data: DataOrLen, erase: bool) {
+        unsafe fn write_flash_inner(&mut self, addr: u32, data: DataOrLen, erase: bool) {
             let (data_ptr, len) = match data {
                 DataOrLen::Data(data) => (data.as_ptr(), data.len()),
                 DataOrLen::Len(len) => (core::ptr::null(), len),
             };
+
+            // this is done on every call in case Self was moved in the mean time.
+            if let Some(boot2) = self.boot2.as_ref() {
+                let boot2_fn_ptr = (boot2 as *const u32 as *const u8).offset(1);
+                let boot2_fn: unsafe extern "C" fn() -> () = core::mem::transmute(boot2_fn_ptr);
+                self.fn_ptrs.flash_enter_cmd_xip = boot2_fn;
+            }
             self.write_flash_inner_in_ram(addr, data_ptr, len, erase)
         }
 
         #[inline(never)]
         #[link_section = ".data.ram_func"]
         unsafe fn write_flash_inner_in_ram(
-            &self,
+            &mut self,
```

Comment:
Why did you add `mut`, here?

Reply:
Mostly to be consistent with the mur on the other method.

The erase/write modify the flash memory that's represented (although the compiler does not know that)